quic/s2n-quic-platform/src/message/cmsg.rs

```diff
@@ -229,88 +304,204 @@ pub fn decode(msghdr: &libc::msghdr) -> AncillaryData {
+        let len = msghdr.msg_controllen as usize;
+        let slice = core::slice::from_raw_parts(ptr, len);
+        Self::new(slice)
+    }
 }
 
 impl<'a> Iterator for Iter<'a> {
-    type Item = &'a libc::cmsghdr;
-
-    fn next(&mut self) -> Option<&'a libc::cmsghdr> {
-        let current = self.cmsghdr.take()?;
-        // CMSG_NXTHDR() returns the next valid cmsghdr after the passed
-        // cmsghdr.  It returns NULL when there isn't enough space left
-        // in the buffer.
-        self.cmsghdr =
-            unsafe { libc::CMSG_NXTHDR(self.msghdr, current).as_ref() }.filter(|cmsghdr| {
-                // CMSG_NXTHDR on some operating systems returns the previous pointer when its
-                // length is zero, so check if the previous pointer is the same as
-                // the current one. Also check if the length is zero, otherwise it'll loop forever.
-                cmsghdr.cmsg_len > 0 && !core::ptr::eq(*cmsghdr, current)
-            });
-        Some(current)
+    type Item = (&'a cmsghdr, &'a [u8]);
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        unsafe {
+            let cursor = self.cursor;
+
+            // make sure we can decode a cmsghdr
+            self.len.checked_sub(size_of::<cmsghdr>())?;
+            let cmsg = &*(cursor as *const cmsghdr);
+            let data_ptr = cursor.add(size_of::<cmsghdr>());
+
+            let cmsg_len = cmsg.cmsg_len as usize;
+
+            // make sure we have capacity to decode the provided cmsg_len
+            self.len.checked_sub(cmsg_len)?;
+
+            // the cmsg_len includes the header itself so it needs to be subtracted off
+            let data_len = cmsg_len.checked_sub(size_of::<cmsghdr>())?;
+            // construct a slice with the provided data len
+            let data = core::slice::from_raw_parts(data_ptr, data_len);
+
+            // empty messages are invalid
+            if data.is_empty() {
+                return None;
+            }
+
+            // calculate the next message and update the cursor/len
+            {
+                let space = libc::CMSG_SPACE(data_len as _) as usize;
+                debug_assert!(
+                    space >= data_len,
+                    "space ({space}) should be at least of size len ({data_len})"
+                );
+                self.len = self.len.saturating_sub(space);
+                self.cursor = cursor.add(space);
+            }
+
+            Some((cmsg, data))
+        }
     }
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
-    use bolero::check;
-    use core::mem::zeroed;
+    use bolero::{check, TypeGenerator};
+    use libc::c_int;
 
+    /// Ensures the cmsg iterator doesn't crash or segfault
     #[test]
-    #[cfg_attr(any(target_os = "linux", target_os = "android"), ignore)] // the linux implementation currently has an integer overflow on garbage data
+    #[cfg_attr(kani, kani::proof, kani::solver(cadical), kani::unwind(17))]
```

Comment:
is cadical the go to now for Kani, or did you try kissat and it was slower?

Reply:
Cadical ended up being faster in this case. I think it just depends on the harness. The advantage cadical has is it's built in process to cbmc, as opposed to kissat that's an external binary. This forces cbmc to write a [cnf](https://people.sc.fsu.edu/~jburkardt/data/cnf/cnf.html) to disk and then pass to kissat. Cadical APIs, instead, can just be called directly.